In [58]:
import pandas as pd
import numpy as np

from urllib.parse import quote    
from urllib.request import urlopen

import sys
import requests
from bs4 import BeautifulSoup 
import pandas as pd
from sqlalchemy import create_engine

#import pymysql.cursors
#pymysql.install_as_MySQLdb()
#import MySQLdb
from pandas.io.json import json_normalize
import time
import datetime
from datetime import timedelta
import re
import folium
from folium.plugins import MarkerCluster, MiniMap

# Read Data

In [2]:
'''bus데이터는 1월자'''
bus = pd.read_csv('BUS_STATION_BOARDING_MONTH_202001.csv',encoding = 'euc-kr')
bus.rename(columns={'역명':'정류장명',
                   '버스정류장ARS번호':'ARSID'},inplace=True)
#bus['ARSID'] = bus['ARSID'].astype('int64') # ARSID에 문자열 정보가 존재

In [4]:
'''지하철 데이터는 12월자'''
subway = pd.read_csv('CARD_SUBWAY_MONTH_201912.csv',encoding = 'euc-kr')

In [4]:
bus_coord = pd.read_csv('서울시버스정류소좌표.csv')

In [5]:
kor_address = pd.read_csv('korea_address.csv')
kor_address.columns = kor_address.iloc[0,:].tolist()
kor_address.drop(index=0,inplace=True)

In [6]:
area_code = pd.read_csv('지역코드.csv')

area_code.columns = ['code','지역구']
pre_code = pd.DataFrame({'code':[1],
              '지역구':['종로구']})

area_code = area_code.append(pre_code)

In [138]:
bus_coord.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11180 entries, 0 to 11179
Data columns (total 5 columns):
ARSID    11180 non-null int64
표준ID     11180 non-null int64
정류장명     11180 non-null object
X좌표      11180 non-null float64
Y좌표      11180 non-null float64
dtypes: float64(2), int64(2), object(1)
memory usage: 436.8+ KB


In [8]:
subway.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18333 entries, 0 to 18332
Data columns (total 7 columns):
사용일자      18333 non-null int64
노선명       18333 non-null object
역ID       18333 non-null int64
역명        18333 non-null object
승차총승객수    18333 non-null int64
하차총승객수    18333 non-null int64
등록일자      18333 non-null int64
dtypes: int64(5), object(2)
memory usage: 1002.7+ KB


In [139]:
#bus arsid 는 문자열이 껴있는 경우가 있음
bus.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1192916 entries, 0 to 1192915
Data columns (total 11 columns):
사용일자         1192916 non-null int64
노선ID         1192916 non-null int64
노선번호         1192916 non-null object
노선명          1192916 non-null object
표준버스정류장ID    1192916 non-null int64
ARSID        1192916 non-null object
역ID          1192916 non-null int64
정류장명         1192916 non-null object
승차총승객수       1192916 non-null int64
하차총승객수       1192916 non-null int64
등록일자         1192916 non-null int64
dtypes: int64(7), object(4)
memory usage: 100.1+ MB


In [19]:
kor_address.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21662 entries, 1 to 21662
Data columns (total 11 columns):
시도            21662 non-null object
시군구           21662 non-null object
행정구역명         21662 non-null object
행정동(행정기관명)    21662 non-null object
법정동           21662 non-null object
행정구역코드        21584 non-null object
행정기관코드        21662 non-null object
행정기관 생성일      21662 non-null object
법정동코드         21662 non-null object
관할지역          1664 non-null object
행정동 영문명칭      21661 non-null object
dtypes: object(11)
memory usage: 2.0+ MB


In [25]:
area_code.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 2 columns):
1      24 non-null int64
종로구    24 non-null object
dtypes: int64(1), object(1)
memory usage: 512.0+ bytes


# subway station name preprocessing

- api로 역사정보를 parsing하기 위해서는 정제된 이름을 경로에 지정해서 넣어줘야 함
- 예를들면 : /서울역 (x) , /서울 (o) -> 오로지 서울역만 뒤에 역명이 붙어서 존재
- 또한 : /잠실(송파구청) (x) , /잠실(0)
- 그리고 현재 api데이터에 우이신설선은 업데이트 되지 않은듯

In [62]:
subway_name0 = subway['역명'].unique().tolist()

station_name = []
for i in subway_name0:
    if '(' in i :
        name_ls0 = i.split('(')
        name = name_ls0[0]
        station_name.append(name)
    else:
        station_name.append(i)

station_name2 = []
for i in station_name:
    if i[-1] == '역':
        station_name2.append(i[:-1])
    else:
        station_name2.append(i)

In [63]:
api_key = '436261446967796537367159686464'
base_url = 'http://swopenAPI.seoul.go.kr/api/subway/{}/xml/stationInfo/0/100/{}'
station_info = pd.DataFrame()
error_num = []
for j in range(len(station_name2)):
    try:
        req = requests.get(base_url.format(api_key,station_name2[j]), headers = {'user-agent': \
                                                ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36')})

        soup = BeautifulSoup(req.content,'html.parser')
        if '해당하는 데이터가 없습니다' in soup.text:
            print(station_name2[j] + ' 수집정보 x')
            error_num.append(j)
            pass
        else:
            
            main_station = []
            statn_id = []
            subway_id = []
            subway_nm = []
            statnf_id = []
            statnf_name = []
            statnt_id = []
            statnt_name = []
            oper = []
            zipno = [] #??
            address_bass = []
            address_detail = []
            statn_tel = []
            x_code = []
            y_code = []
            transit = []

            for_loop = len(soup.find_all('statnnm'))

            for i in range(for_loop):
                main_station.append(soup.find_all('statnnm')[i].text)
                statn_id.append(soup.find_all('statnid')[i].text)
                subway_id.append(soup.find_all('subwayid')[i].text)
                subway_nm.append(soup.find_all('subwaynm')[i].text)
                statnf_id.append(soup.find_all('statnfid')[i].text)
                statnf_name.append(soup.find_all('statnfnm')[i].text)
                statnt_id.append(soup.find_all('statntid')[i].text)
                statnt_name.append(soup.find_all('statntnm')[i].text)
                oper.append(soup.find_all('operpblinstt')[i].text)
                zipno.append(soup.find_all('zipno')[i].text)
                address_bass.append(soup.find_all('adresbass')[i].text)
                address_detail.append(soup.find_all('adresdetail')[i].text)
                statn_tel.append(soup.find_all('telno')[i].text)
                x_code.append(soup.find_all('subwayxcnts')[i].text)
                y_code.append(soup.find_all('subwayycnts')[i].text)
                transit.append(soup.find_all('trnsitco')[i].text)

            station_info0 = pd.DataFrame({'statn_id' : statn_id,
                                          'main_station' : main_station,
                                          'subway_id' : subway_id,
                                          'subway_nm' : subway_nm,
                                          'statnf_id' : statnf_id,
                                          'statnf_name' : statnf_name,
                                          'statnt_id' : statnt_id,
                                          'statnt_name' : statnt_name,
                                          'oper' : oper,
                                          'zipno' : zipno,
                                          'address_bass' : address_bass,
                                          'address_detail' : address_detail,
                                          'x_code' : x_code,
                                          'y_code' : y_code,
                                          'transit' : transit})

            station_info = station_info.append(station_info0)
            print('--------------'+ station_name2[j] + ' 역 완료----------------')
    except:
        print(station_name2[j] + ' 역 request error')
        pass

--------------종각 역 완료----------------
--------------종로3가 역 완료----------------
--------------종로5가 역 완료----------------
--------------동대문 역 완료----------------
--------------신설동 역 완료----------------
--------------제기동 역 완료----------------
--------------청량리 역 완료----------------
--------------동묘앞 역 완료----------------
--------------시청 역 완료----------------
--------------을지로입구 역 완료----------------
--------------을지로3가 역 완료----------------
--------------을지로4가 역 완료----------------
--------------동대문역사문화공원 역 완료----------------
--------------신당 역 완료----------------
--------------상왕십리 역 완료----------------
--------------왕십리 역 완료----------------
--------------한양대 역 완료----------------
--------------뚝섬 역 완료----------------
--------------성수 역 완료----------------
--------------건대입구 역 완료----------------
--------------구의 역 완료----------------
--------------강변 역 완료----------------
--------------잠실나루 역 완료----------------
--------------잠실 역 완료----------------
--------------잠실새내 역 완료----------------
--------------종

--------------개포동 역 완료----------------
--------------대모산입구 역 완료----------------
--------------서울숲 역 완료----------------
--------------압구정로데오 역 완료----------------
--------------강남구청 역 완료----------------
--------------선정릉 역 완료----------------
--------------가천대 역 완료----------------
--------------태평 역 완료----------------
--------------모란 역 완료----------------
--------------야탑 역 완료----------------
--------------서현 역 완료----------------
--------------수내 역 완료----------------
--------------정자 역 완료----------------
--------------미금 역 완료----------------
--------------오리 역 완료----------------
--------------이매 역 완료----------------
--------------보정 역 완료----------------
--------------죽전 역 완료----------------
--------------구성 역 완료----------------
--------------신갈 역 완료----------------
--------------기흥 역 완료----------------
--------------상갈 역 완료----------------
--------------청명 역 완료----------------
--------------영통 역 완료----------------
--------------망포 역 완료----------------
--------------매탄권선 역 완료--------------

--------------보라매 역 완료----------------
--------------신풍 역 완료----------------
--------------남구로 역 완료----------------
--------------철산 역 완료----------------
--------------광명사거리 역 완료----------------
--------------천왕 역 완료----------------
--------------까치울 역 완료----------------
--------------부천종합운동장 역 완료----------------
--------------춘의 역 완료----------------
--------------신중동 역 완료----------------
--------------부천시청 역 완료----------------
--------------상동 역 완료----------------
--------------삼산체육관 역 완료----------------
--------------굴포천 역 완료----------------
--------------부평구청 역 완료----------------
--------------암사 역 완료----------------
--------------강동구청 역 완료----------------
--------------몽촌토성 역 완료----------------
--------------석촌 역 완료----------------
--------------송파 역 완료----------------
--------------문정 역 완료----------------
--------------장지 역 완료----------------
--------------복정 역 완료----------------
--------------산성 역 완료----------------
--------------남한산성입구 역 완료----------------
--------------단대오거리 역 

In [71]:
station_info.to_csv('station_info.csv')
station_info.head()

statn_id main_station subway_id subway_nm   statnf_id statnf_name  \
0  1001000131           종각      1001       1호선  1001000130        종로3가   
0  1003000329         종로3가      1003       3호선  1003000328          안국   
1  1005000534         종로3가      1005       5호선  1005000533         광화문   
2  1001000130         종로3가      1001       1호선  1001000129        종로5가   
0  1001000129         종로5가      1001       1호선  1001000128         동대문   

    statnt_id statnt_name   oper   zipno    address_bass address_detail  \
0  1001000132          시청  서울메트로  110121  서울특별시 종로구 종로1가             44   
0  1003000330       을지로3가  서울메트로  110330   서울특별시 종로구 돈의동             46   
1  1005000535       을지로4가   도시철도  110330   서울특별시 종로구 돈의동             46   
2  1001000131          종각  서울메트로  110330   서울특별시 종로구 돈의동             46   
0  1001000130        종로3가  서울메트로  110125  서울특별시 종로구 종로5가          124-8   

        x_code       y_code transit  
0  198520.9972  452310.4761       1  
0     199270.9     452469.7       3  
1  199127.4283  452568.6269       3  
2  199296.4693   452334.288       3  
0  200171.3944  452385.8684       1

In [65]:
#데이터상으로 아직 우이신설선 데이터 x
for i in error_num:
    print(station_name2[i])

총신대입구
쌍용
삼동
경기광주
초월
곤지암
신둔도예촌
이천
부발
세종대왕릉
여주
이수
북한산우이
솔밭공원
4.19민주묘지
가오리
화계
삼양
삼양사거리
솔샘
북한산보국문
정릉


In [66]:
# station 구 추출
'''
각 시별로 구의 위치가 다르고 또한 구가 없는 지역도 있으므로 해당 지역들은 따로 처리를 하고
서울 지역의 구만 추출
'''
for i in range(len(station_info)):
    if len(station_info['address_bass'].iloc[i].split(' ')) != 3:
        print(station_info['address_bass'].iloc[i])
    else:
        pass

강북구 우이동
서울특별시 동대문구 왕산로 214
서울특별시 동대문구 왕산로 214
서울특별시 동대문구 왕산로 214
서울특별시 성동구 왕십리광장로 17
서울특별시 성동구 완십리광장로 17
서울특별시 강남구 강남대로 지하396
서울특별시 마포구 양화로 지하188
경기도 고양시 덕양구 지축동
서울특별시 성동구 동호로 지하21
서울특별시 서초구 남부순환로 지하2585
성북구 동소문동
서울특별시 성북구 동선동 4가
서울특별시 용산구 한강대로 405
서울특별시 용산구 이촌로71길 42
서울특별시 용산구 한강대로23길 55
경기도 안양시 만안구 석수동
경기도 안양시 만안구 안양1동
경기도 안양시 만안구 안양6동
경기도 수원시 장안구 율전동
경기도 수원시 장안구 화서2동
경기도 수원시 팔달구 매산로1가 18
경기도 수원시 권선구 매산로1가
경기도 수원시 권선구 장지동
경기도 평택시 진위면 하북리
충청남도 천안시 성환읍 성환리
충청남도 천안시 직산읍 모시리
경기도 부천시 원미구 역곡동
경기도 부천시 소사구 심곡본동
경기도 부천시 소사구 송내동
인천광역시 부평구 광장로 지하15
인천광역시 남구 주안동 125
인천광역시 중구 북성동 2가 14
경기도 부천시 소사구 소사본2동
경기도 부천시 소사구 송내동
서울특별시 용산구 서빙고로 238
서울특별시 용산구 독서당로6길 12-13
서울특별시 성동구 고산자로 123
서울특별시 노원구 석계로 98-2
경기도 의정부시 평화로 363
경기도 안산시 상록구 대야미동
경기도 안산시 상록구 건건동
경기도 안산시 상록구 일동
경기도 안산시 상록구 이동
경기도 안산시 단원구 고잔동
경기도 안산시 단원구 고잔동
경기도 안산시 단원구 초지동
경기도 안산시 단원구 초지동
경기도 안산시 단원구 원곡동
경기도 안산시 단원구 신길동
경기도 시흥시 역전로 430
경기도 안양시 동안구 관양2동
경기도 안양시 동안구 부림동
경기도 안앙시 동안구 호계동
서울특별시 성동구 왕십리로 지하77
서울특별시 강남구 압구정로 지하402
서울특별시 강남구 학동로 지하346
서울

In [69]:
ku_ls = []
for i in range(len(station_info)):
    if '서울' not in station_info['address_bass'].iloc[i]:
        ku_ls.append(np.nan)
    else:
        ku_ls.append(station_info['address_bass'].iloc[i].split(' ')[1])
station_info['ku'] = ku_ls

# subway 월별 시간대별 데이터

In [2]:
engine = create_engine("mysql+mysqldb://justkaggle:"+"justkaggle1234!"+"@104.197.128.149/smy_db", encoding='utf-8')
conn = engine.connect()

In [3]:
api_key = '4c6e727a6c6779653532546b576f6e'
base_url = 'http://openapi.seoul.go.kr:8088/{}/json/CardSubwayTime/{}/{}/{}/'

year_range = ['2015','2016','2017','2018','2019']
month_range = ['01','02','03','04','05','06','07','08','09','10','11','12']

for j in year_range:
    start_time = time.time()
    table_name = 'subway_timelog_'+j
    
    for i in month_range:
        ym_value = j + i
        
        num = 1
        while True:
            try:

                req = requests.get(base_url.format(api_key,str(num),str(num+999),ym_value), headers = {'user-agent': \
                                                ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36')})

                subway_json = req.json()
                sample_df = json_normalize(subway_json['CardSubwayTime']['row'])

                sample_df.to_sql(name=table_name,con=engine,if_exists='append')


                print('지하철 데이터 {} ~ {} row 수집중'.format(str(num),str(num+999)))
                num +=1000
            except:
                break
        print('--------------------'+j+'년'+i+'월 데이터 수집 완료---------------------')
        
        
    print('============================================================================')
    print('=============================='+j+' 연도 테이블 생성 완료========================')
    print('============================================================================')
    print('연평균 수집 소요 시간(초) : ', time.time()-start_time)



지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년01월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년02월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년03월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년04월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년05월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년06월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년07월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년08월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년09월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년10월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년11월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2015년12월 데이터 수집 완료--

In [22]:
api_key = '4c6e727a6c6779653532546b576f6e'
base_url = 'http://openapi.seoul.go.kr:8088/{}/json/CardSubwayTime/{}/{}/{}/'

year_range = ['2020']
month_range = ['01','02','03','04','05','06','07','08','09','10','11','12']

start_time = time.time()
sub_ym_df = pd.DataFrame()
for j in year_range:
    
    for i in month_range:
    
        ym_value = j + i
            
        if datetime.datetime.strptime(ym_value,'%Y%M') < datetime.datetime.strptime('202008','%Y%M'):
            pass
        else:
            print('데이터 존재 X')
            continue
        
        sub_ym_df0 = pd.DataFrame()
        num = 1
        while True:
            try:

                req = requests.get(base_url.format(api_key,str(num),str(num+999),ym_value), headers = {'user-agent': \
                                                ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36')})

                subway_json = req.json()
                sample_df = json_normalize(subway_json['CardSubwayTime']['row'])

                #sample_df.to_sql(name=table_name,con=engine,if_exists='append')
                sub_ym_df0 = sub_ym_df0.append(sample_df)

                print('지하철 데이터 {} ~ {} row 수집중'.format(str(num),str(num+999)))
                num +=1000
            except:
                break
        print('--------------------'+j+'년'+i+'월 데이터 수집 완료---------------------')
        
        #sub_ym_df0['ym'] = ym_value
        sub_ym_df = sub_ym_df.append(sub_ym_df0)
        
    print('============================================================================')
    print('=============================='+j+' 연도 테이블 생성 완료========================')
    print('============================================================================')
    print('연평균 수집 소요 시간(초) : ', time.time()-start_time)


/Users/sinmin-yong/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:30: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


지하철 데이터 1 ~ 1000 row 수집중
--------------------2020년01월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2020년02월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2020년03월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2020년04월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2020년05월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2020년06월 데이터 수집 완료---------------------
지하철 데이터 1 ~ 1000 row 수집중
--------------------2020년07월 데이터 수집 완료---------------------
데이터 존재 X
데이터 존재 X
데이터 존재 X
데이터 존재 X
데이터 존재 X
==============================2020 연도 테이블 생성 완료========================
연평균 수집 소요 시간(초) :  2.657386064529419


In [24]:
sub_ym_df.head()

USE_MON LINE_NUM SUB_STA_NM  FOUR_RIDE_NUM  FOUR_ALIGHT_NUM  FIVE_RIDE_NUM  \
0  202001      8호선         산성           22.0              0.0         4469.0   
1  202001      1호선        서울역          908.0             18.0         8948.0   
2  202001      1호선         시청           55.0              0.0         1999.0   
3  202001      1호선         종각          126.0              5.0         4380.0   
4  202001      1호선       종로3가          160.0             25.0         4750.0   

   FIVE_ALIGHT_NUM  SIX_RIDE_NUM  SIX_ALIGHT_NUM  SEVEN_RIDE_NUM  ...  \
0            865.0        7308.0          5246.0         17343.0  ...   
1           7940.0       13146.0         47332.0         37251.0  ...   
2           3976.0        3497.0         19177.0          7417.0  ...   
3           4683.0        4456.0         25858.0          6286.0  ...   
4           2452.0        4234.0         13222.0          5495.0  ...   

   TWENTY_THREE_ALIGHT_NUM  MIDNIGHT_RIDE_NUM  MIDNIGHT_ALIGHT_NUM  \
0                   5877.0              162.0               1885.0   
1                  17549.0             4190.0               5633.0   
2                   3747.0              556.0               1063.0   
3                   5038.0             2896.0               1577.0   
4                   5357.0             2565.0               2284.0   

   ONE_RIDE_NUM  ONE_ALIGHT_NUM  TWO_RIDE_NUM  TWO_ALIGHT_NUM  THREE_RIDE_NUM  \
0           1.0             8.0           0.0             0.0             0.0   
1          98.0           291.0           0.0             8.0             0.0   
2           7.0            38.0           0.0             0.0             0.0   
3           9.0            22.0           0.0             0.0             0.0   
4          30.0            89.0           0.0             0.0             0.0   

   THREE_ALIGHT_NUM   WORK_DT  
0               0.0  20200203  
1               0.0  20200203  
2               0.0  20200203  
3               0.0  20200203  
4               0.0  20200203  

[5 rows x 52 columns]

# subway 일별 데이터

In [11]:
engine = create_engine("mysql+mysqldb://justkaggle:"+"justkaggle1234!"+"@104.197.128.149/smy_db", encoding='utf-8')
conn = engine.connect()

In [12]:
api_key = '7a4d4a55776779653838644e786969'
base_url = 'http://openapi.seoul.go.kr:8088/{}/json/CardSubwayStatsNew/{}/{}/{}'

year_range = ['2015','2016','2017','2018','2019']
date_range = pd.date_range(start='20150101', end='20191231').strftime('%Y%m%d').tolist()
month_range = []
for i in date_range:
    month_range.append(i[4:])
day_range = pd.Series(month_range).unique().tolist()[:-1]   

In [14]:
for i in year_range:
    start_time = time.time()
    db_name = 'subway_dailylog_' + i
    
    for j in day_range:
        ym = i+j
        
        num = 1
        
        while True:
            try:
                req = requests.get(base_url.format(api_key,str(num),str(num+999),ym), headers = {'user-agent': \
                                                        ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36')})

                json_data = req.json()
                sample_df = json_normalize(json_data['CardSubwayStatsNew']['row'])
                
                sample_df.to_sql(name=db_name,con=engine,if_exists='append')
                
                #print('지하철 데이터 {} ~ {} row 수집중'.format(str(num),str(num+999)))
                print(i+j+' 데이터 적재 완료')
                num +=1000
            except:
                break
                
    print('============================================================================')
    print('=============================='+i+' 연도 테이블 생성 완료========================')
    print('============================================================================')
    print('연평균 수집 소요 시간(초) : ', time.time()-start_time)

        

20150101 데이터 적재 완료
20150102 데이터 적재 완료
20150103 데이터 적재 완료
20150104 데이터 적재 완료
20150105 데이터 적재 완료
20150106 데이터 적재 완료
20150107 데이터 적재 완료
20150108 데이터 적재 완료
20150109 데이터 적재 완료
20150110 데이터 적재 완료
20150111 데이터 적재 완료
20150112 데이터 적재 완료
20150113 데이터 적재 완료
20150114 데이터 적재 완료
20150115 데이터 적재 완료
20150116 데이터 적재 완료
20150117 데이터 적재 완료
20150118 데이터 적재 완료
20150119 데이터 적재 완료
20150120 데이터 적재 완료
20150121 데이터 적재 완료
20150122 데이터 적재 완료
20150123 데이터 적재 완료
20150124 데이터 적재 완료
20150125 데이터 적재 완료
20150126 데이터 적재 완료
20150127 데이터 적재 완료
20150128 데이터 적재 완료
20150129 데이터 적재 완료
20150130 데이터 적재 완료
20150131 데이터 적재 완료
20150201 데이터 적재 완료
20150202 데이터 적재 완료
20150203 데이터 적재 완료
20150204 데이터 적재 완료
20150205 데이터 적재 완료
20150206 데이터 적재 완료
20150207 데이터 적재 완료
20150208 데이터 적재 완료
20150209 데이터 적재 완료
20150210 데이터 적재 완료
20150211 데이터 적재 완료
20150212 데이터 적재 완료
20150213 데이터 적재 완료
20150214 데이터 적재 완료
20150215 데이터 적재 완료
20150216 데이터 적재 완료
20150217 데이터 적재 완료
20150218 데이터 적재 완료
20150219 데이터 적재 완료
20150220 데이터 적재 완료
20150221 데이터 적재 완료
20150222 데이터

20160223 데이터 적재 완료
20160224 데이터 적재 완료
20160225 데이터 적재 완료
20160226 데이터 적재 완료
20160227 데이터 적재 완료
20160228 데이터 적재 완료
20160301 데이터 적재 완료
20160302 데이터 적재 완료
20160303 데이터 적재 완료
20160304 데이터 적재 완료
20160305 데이터 적재 완료
20160306 데이터 적재 완료
20160307 데이터 적재 완료
20160308 데이터 적재 완료
20160309 데이터 적재 완료
20160310 데이터 적재 완료
20160311 데이터 적재 완료
20160312 데이터 적재 완료
20160313 데이터 적재 완료
20160314 데이터 적재 완료
20160315 데이터 적재 완료
20160316 데이터 적재 완료
20160317 데이터 적재 완료
20160318 데이터 적재 완료
20160319 데이터 적재 완료
20160320 데이터 적재 완료
20160321 데이터 적재 완료
20160322 데이터 적재 완료
20160323 데이터 적재 완료
20160324 데이터 적재 완료
20160325 데이터 적재 완료
20160326 데이터 적재 완료
20160327 데이터 적재 완료
20160328 데이터 적재 완료
20160329 데이터 적재 완료
20160330 데이터 적재 완료
20160331 데이터 적재 완료
20160401 데이터 적재 완료
20160402 데이터 적재 완료
20160403 데이터 적재 완료
20160404 데이터 적재 완료
20160405 데이터 적재 완료
20160406 데이터 적재 완료
20160407 데이터 적재 완료
20160408 데이터 적재 완료
20160409 데이터 적재 완료
20160410 데이터 적재 완료
20160411 데이터 적재 완료
20160412 데이터 적재 완료
20160413 데이터 적재 완료
20160414 데이터 적재 완료
20160415 데이터 적재 완료
20160416 데이터

20170417 데이터 적재 완료
20170418 데이터 적재 완료
20170419 데이터 적재 완료
20170420 데이터 적재 완료
20170421 데이터 적재 완료
20170422 데이터 적재 완료
20170423 데이터 적재 완료
20170424 데이터 적재 완료
20170425 데이터 적재 완료
20170426 데이터 적재 완료
20170427 데이터 적재 완료
20170428 데이터 적재 완료
20170429 데이터 적재 완료
20170430 데이터 적재 완료
20170501 데이터 적재 완료
20170502 데이터 적재 완료
20170503 데이터 적재 완료
20170504 데이터 적재 완료
20170505 데이터 적재 완료
20170506 데이터 적재 완료
20170507 데이터 적재 완료
20170508 데이터 적재 완료
20170509 데이터 적재 완료
20170510 데이터 적재 완료
20170511 데이터 적재 완료
20170512 데이터 적재 완료
20170513 데이터 적재 완료
20170514 데이터 적재 완료
20170515 데이터 적재 완료
20170516 데이터 적재 완료
20170517 데이터 적재 완료
20170518 데이터 적재 완료
20170519 데이터 적재 완료
20170520 데이터 적재 완료
20170521 데이터 적재 완료
20170522 데이터 적재 완료
20170523 데이터 적재 완료
20170524 데이터 적재 완료
20170525 데이터 적재 완료
20170526 데이터 적재 완료
20170527 데이터 적재 완료
20170528 데이터 적재 완료
20170529 데이터 적재 완료
20170530 데이터 적재 완료
20170531 데이터 적재 완료
20170601 데이터 적재 완료
20170602 데이터 적재 완료
20170603 데이터 적재 완료
20170604 데이터 적재 완료
20170605 데이터 적재 완료
20170606 데이터 적재 완료
20170607 데이터 적재 완료
20170608 데이터

20180609 데이터 적재 완료
20180610 데이터 적재 완료
20180611 데이터 적재 완료
20180612 데이터 적재 완료
20180613 데이터 적재 완료
20180614 데이터 적재 완료
20180615 데이터 적재 완료
20180616 데이터 적재 완료
20180617 데이터 적재 완료
20180618 데이터 적재 완료
20180619 데이터 적재 완료
20180620 데이터 적재 완료
20180621 데이터 적재 완료
20180622 데이터 적재 완료
20180623 데이터 적재 완료
20180624 데이터 적재 완료
20180625 데이터 적재 완료
20180626 데이터 적재 완료
20180627 데이터 적재 완료
20180628 데이터 적재 완료
20180629 데이터 적재 완료
20180630 데이터 적재 완료
20180701 데이터 적재 완료
20180702 데이터 적재 완료
20180703 데이터 적재 완료
20180704 데이터 적재 완료
20180705 데이터 적재 완료
20180706 데이터 적재 완료
20180707 데이터 적재 완료
20180708 데이터 적재 완료
20180709 데이터 적재 완료
20180710 데이터 적재 완료
20180711 데이터 적재 완료
20180712 데이터 적재 완료
20180713 데이터 적재 완료
20180714 데이터 적재 완료
20180715 데이터 적재 완료
20180716 데이터 적재 완료
20180717 데이터 적재 완료
20180718 데이터 적재 완료
20180719 데이터 적재 완료
20180720 데이터 적재 완료
20180721 데이터 적재 완료
20180722 데이터 적재 완료
20180723 데이터 적재 완료
20180724 데이터 적재 완료
20180725 데이터 적재 완료
20180726 데이터 적재 완료
20180727 데이터 적재 완료
20180728 데이터 적재 완료
20180729 데이터 적재 완료
20180730 데이터 적재 완료
20180731 데이터

20190801 데이터 적재 완료
20190802 데이터 적재 완료
20190803 데이터 적재 완료
20190804 데이터 적재 완료
20190805 데이터 적재 완료
20190806 데이터 적재 완료
20190807 데이터 적재 완료
20190808 데이터 적재 완료
20190809 데이터 적재 완료
20190810 데이터 적재 완료
20190811 데이터 적재 완료
20190812 데이터 적재 완료
20190813 데이터 적재 완료
20190814 데이터 적재 완료
20190815 데이터 적재 완료
20190816 데이터 적재 완료
20190817 데이터 적재 완료
20190818 데이터 적재 완료
20190819 데이터 적재 완료
20190820 데이터 적재 완료
20190821 데이터 적재 완료
20190822 데이터 적재 완료
20190823 데이터 적재 완료
20190824 데이터 적재 완료
20190825 데이터 적재 완료
20190826 데이터 적재 완료
20190827 데이터 적재 완료
20190828 데이터 적재 완료
20190829 데이터 적재 완료
20190830 데이터 적재 완료
20190831 데이터 적재 완료
20190901 데이터 적재 완료
20190902 데이터 적재 완료
20190903 데이터 적재 완료
20190904 데이터 적재 완료
20190905 데이터 적재 완료
20190906 데이터 적재 완료
20190907 데이터 적재 완료
20190908 데이터 적재 완료
20190909 데이터 적재 완료
20190910 데이터 적재 완료
20190911 데이터 적재 완료
20190912 데이터 적재 완료
20190913 데이터 적재 완료
20190914 데이터 적재 완료
20190915 데이터 적재 완료
20190916 데이터 적재 완료
20190917 데이터 적재 완료
20190918 데이터 적재 완료
20190919 데이터 적재 완료
20190920 데이터 적재 완료
20190921 데이터 적재 완료
20190922 데이터

# bus preprocessing

In [12]:
bus

사용일자      노선ID  노선번호                   노선명  표준버스정류장ID  ARSID  \
0        20200101  11110371   N62  N62번(양천공영차고지~면목동차고지)  103000034  04133   
1        20200101  41110061  6616         6616(철산동~온수동)  116000120  17210   
2        20200101  41110061  6616         6616(철산동~온수동)  116000080  17166   
3        20200101  41110061  6616         6616(철산동~온수동)  116000083  17169   
4        20200101  41110061  6616         6616(철산동~온수동)  116000085  17171   
...           ...       ...   ...                   ...        ...    ...   
1192911  20200131  11111163  강남10    강남10(개포4단지7단지~양재역)  122000127  23230   
1192912  20200131  11111163  강남10    강남10(개포4단지7단지~양재역)  122000128  23231   
1192913  20200131  11111163  강남10    강남10(개포4단지7단지~양재역)  122000360  23477   
1192914  20200131  11111163  강남10    강남10(개포4단지7단지~양재역)  122000358  23475   
1192915  20200131  11111163  강남10    강남10(개포4단지7단지~양재역)  122000359  23476   

             역ID             정류장명  승차총승객수  하차총승객수      등록일자  
0          72645             왕십리역       0      13  20200104  
1          10202     현대아파트.개봉중앙시장     153      92  20200104  
2        8001850            경인중학교     171      46  20200104  
3        9004386  오류1동주민센터.오류문화센터      87      16  20200104  
4        9004384             오류동역     129      46  20200104  
...          ...              ...     ...     ...       ...  
1192911  9102654          대치역8번출구      86     130  20200203  
1192912  9000159          대치역7번출구      16      36  20200203  
1192913  9033019         개포4단지7단지       4      33  20200203  
1192914  9033024         개포4단지5단지       0      39  20200203  
1192915  9033022         개포4단지6단지       0      36  20200203  

[1192916 rows x 11 columns]

In [77]:
pd.merge(bus,bus_coord, on='정류장명',how='left').head()

사용일자      노선ID  노선번호                   노선명  표준버스정류장ID ARSID_x    역ID  \
0  20200101  11110371   N62  N62번(양천공영차고지~면목동차고지)  103000034   04133  72645   
1  20200101  11110371   N62  N62번(양천공영차고지~면목동차고지)  103000034   04133  72645   
2  20200101  11110371   N62  N62번(양천공영차고지~면목동차고지)  103000034   04133  72645   
3  20200101  11110371   N62  N62번(양천공영차고지~면목동차고지)  103000034   04133  72645   
4  20200101  41110061  6616         6616(철산동~온수동)  116000120   17210  10202   

           정류장명  승차총승객수  하차총승객수      등록일자  ARSID_y         표준ID         X좌표  \
0          왕십리역       0      13  20200104   4133.0  103000034.0  127.036269   
1          왕십리역       0      13  20200104   4134.0  103000035.0  127.036193   
2          왕십리역       0      13  20200104   4135.0  103000036.0  127.034901   
3          왕십리역       0      13  20200104   4781.0  105000472.0  127.037685   
4  현대아파트.개봉중앙시장     153      92  20200104  17209.0  116000119.0  126.856039   

         Y좌표  
0  37.559988  
1  37.560298  
2  37.560770  
3  37.560936  
4  37.490892

In [75]:
bus_coord.head()

ARSID       표준ID           정류장명         X좌표        Y좌표
0   1001  100000001        종로2가사거리  126.987786  37.569764
1   1002  100000002    창경궁.서울대학교병원  126.996520  37.579179
2   1003  100000003      명륜3가.성대입구  126.998290  37.582709
3   1004  100000004       종로2가.삼일교  126.987507  37.568582
4   1005  100000005  혜화동로터리.여운형활동터  127.001694  37.586230

In [36]:
bus.head()

사용일자      노선ID  노선번호                   노선명  표준버스정류장ID  ARSID      역ID  \
0  20200101  11110371   N62  N62번(양천공영차고지~면목동차고지)  103000034  04133    72645   
1  20200101  41110061  6616         6616(철산동~온수동)  116000120  17210    10202   
2  20200101  41110061  6616         6616(철산동~온수동)  116000080  17166  8001850   
3  20200101  41110061  6616         6616(철산동~온수동)  116000083  17169  9004386   
4  20200101  41110061  6616         6616(철산동~온수동)  116000085  17171  9004384   

              정류장명  승차총승객수  하차총승객수      등록일자  
0             왕십리역       0      13  20200104  
1     현대아파트.개봉중앙시장     153      92  20200104  
2            경인중학교     171      46  20200104  
3  오류1동주민센터.오류문화센터      87      16  20200104  
4             오류동역     129      46  20200104

In [ ]:
'''표준버스정류장id에 대한 분석이 필요 (서울의 데이터만 필요하므로)'''

# bus coord to address

In [ ]:
'''bus coord데이터에 있는 표준버스정류장id의 맨 앞자리가 1인 것을 확인 -> 모두 서울'''
for i in range(len(bus_coord)):
    
    if str(bus_coord['표준ID'].iloc[i])[:1] == '1':
        continue
    else:
        print(bus_coord['정류장명'].iloc[i])
        

In [45]:
'''coord to address api 활용'''



In [72]:
kor_address.head()

시도    시군구  행정구역명 행정동(행정기관명)    법정동   행정구역코드      행정기관코드  행정기관 생성일  \
1  서울특별시  서울특별시  서울특별시      서울특별시  서울특별시       11  1100000000  19880423   
2  서울특별시    종로구    종로구        종로구    종로구    11010  1111000000  19880423   
3  서울특별시    종로구  청운효자동      청운효자동    청운동  1101072  1111051500  20081101   
4  서울특별시    종로구  청운효자동      청운효자동    신교동  1101072  1111051500  20081101   
5  서울특별시    종로구  청운효자동      청운효자동    궁정동  1101072  1111051500  20081101   

        법정동코드 관할지역            행정동 영문명칭  
1  1100000000  NaN               Seoul  
2  1111000000  NaN           Jongno-gu  
3  1111010100  NaN  Cheongunhyoja-dong  
4  1111010200  NaN  Cheongunhyoja-dong  
5  1111010300  NaN  Cheongunhyoja-dong

In [73]:
area_code.head()

code   지역구
0     2    중구
1     3   용산구
2     4   성동구
3     5   광진구
4     6  동대문구

# 버스 일별 데이터 api -> db
- 현재는 dataframe을 만들어 저장하는 구조인데, 이후에는 각 행을 리스트로 만들어서 각 행별로 바로 db에 넣는 작업 실시 예정

In [2]:
#just kaggle
engine = create_engine("mysql+mysqldb://justkaggle:"+"justkaggle1234!"+"@104.197.128.149/smy_db", encoding='utf-8')
conn = engine.connect()

year_range = ['2015','2016','2017','2018','2019']
date_range = pd.date_range(start='20150101', end='20191231').strftime('%Y%m%d').tolist()
month_range = []
for i in date_range:
    month_range.append(i[4:])

api_key = '5a7471586f6779653634776c4a4952'
base_url = 'http://openapi.seoul.go.kr:8088/{}/json/CardBusStatisticsServiceNew/{}/{}/{}'
#bus_info = pd.DataFrame()
#error_num = []


In [ ]:
for j in year_range:
    start_time = time.time()
    db_name = 'bus_daylog_'+j
    for i in month_range:
        date_time = j+i
        num = 1
        while True:
            try:

                req = requests.get(base_url.format(api_key,str(num),str(num+999),date_time), headers = {'user-agent': \
                                                ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36')})

                bus_json = req.json()
                sample_df = json_normalize(bus_json['CardBusStatisticsServiceNew']['row'])
                
                sample_df.to_sql(name=db_name,con = engine,if_exists='append')
                

                num +=1000
            except:
                break
        print('--------------------'+i+' 일자 데이터 수집 완료---------------------')
        print('--------------------'+str(num)+'행 수집-------------------------')
        
        
    print('============================================================================')
    print('=============================='+j+' 연도 테이블 생성 완료========================')
    print('============================================================================')
    print('연평균 수집 소요 시간(초) : ', time.time()-start_time)
    #평균 8시간정도 소요

--------------------0101 일자 데이터 수집 완료---------------------
--------------------37001행 수집-------------------------
--------------------0102 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0103 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0104 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0105 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0106 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0107 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0108 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0109 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------0314 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0315 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0316 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0317 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0318 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0319 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0320 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0321 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0322 일자 데이터 수집 완료---------------------
--------------------37001행 수집

--------------------0525 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0526 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0527 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0528 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0529 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0530 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0531 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0601 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0602 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------0805 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0806 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0807 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0808 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0809 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0810 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0811 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0812 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0813 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------1016 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1017 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1018 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1019 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1020 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1021 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1022 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1023 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1024 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------1227 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1228 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1229 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1230 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1231 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0101 일자 데이터 수집 완료---------------------
--------------------37001행 수집-------------------------
--------------------0102 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0103 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0104 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------0308 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0309 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0310 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0311 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0312 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0313 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0314 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0315 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0316 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------0519 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0520 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0521 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0522 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0523 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0524 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0525 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0526 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0527 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------0730 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0731 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0801 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0802 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0803 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0804 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0805 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0806 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------0807 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------1010 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1011 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1012 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1013 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1014 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1015 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1016 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1017 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1018 일자 데이터 수집 완료---------------------
--------------------38001행 수집

--------------------1221 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1222 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1223 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1224 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1225 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1226 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1227 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1228 일자 데이터 수집 완료---------------------
--------------------38001행 수집-------------------------
--------------------1229 일자 데이터 수집 완료---------------------
--------------------38001행 수집

In [220]:
bus_info.to_csv('bus_201501.csv',index=False)

# 버스 시간대별 데이터 수집 코드
- 연도별로 데이터 삽입
- 1000행 씩 insert

In [288]:
import pymysql

In [292]:
conn = pymysql.connect(host='104.197.128.149', user='justkaggle', password='justkaggle1234!',
                           db='smy_db', charset='utf8')
curs = conn.cursor()

In [295]:
sql = '''INSERT INTO bus_timelog_201901(
USE_MON, BUS_ROUTE_NO, BUS_ROUTE_NM, STND_BSST_ID,
       BSST_ARS_NO, BUS_STA_NM, MIDNIGHT_RIDE_NUM, MIDNIGHT_ALIGHT_NUM,
       ONE_RIDE_NUM, ONE_ALIGHT_NUM, TWO_RIDE_NUM, TWO_ALIGHT_NUM,
       THREE_RIDE_NUM, THREE_ALIGHT_NUM, FOUR_RIDE_NUM,
       FOUR_ALIGHT_NUM, FIVE_RIDE_NUM, FIVE_ALIGHT_NUM, SIX_RIDE_NUM,
       SIX_ALIGHT_NUM, SEVEN_RIDE_NUM, SEVEN_ALIGHT_NUM,
       EIGHT_RIDE_NUM, EIGHT_ALIGHT_NUM, NINE_RIDE_NUM,
       NINE_ALIGHT_NUM, TEN_RIDE_NUM, TEN_ALIGHT_NUM, ELEVEN_RIDE_NUM,
       ELEVEN_ALIGHT_NUM, TWELVE_RIDE_NUM, TWELVE_ALIGHT_NUM,
       THIRTEEN_RIDE_NUM, THIRTEEN_ALIGHT_NUM, FOURTEEN_RIDE_NUM,
       FOURTEEN_ALIGHT_NUM, FIFTEEN_RIDE_NUM, FIFTEEN_ALIGHT_NUM,
       SIXTEEN_RIDE_NUM, SIXTEEN_ALIGHT_NUM, SEVENTEEN_RIDE_NUM,
       SEVENTEEN_ALIGHT_NUM, EIGHTEEN_RIDE_NUM, EIGHTEEN_ALIGHT_NUM,
       NINETEEN_RIDE_NUM, NINETEEN_ALIGHT_NUM, TWENTY_RIDE_NUM,
       TWENTY_ALIGHT_NUM, TWENTY_ONE_RIDE_NUM, TWENTY_ONE_ALIGHT_NUM,
       TWENTY_TWO_RIDE_NUM, TWENTY_TWO_ALIGHT_NUM, TWENTY_THREE_RIDE_NUM,
       TWENTY_THREE_ALIGHT_NUM, WORK_DT) 
       values(%s,%s,%s,

In [25]:
#just kaggle
#engine = create_engine("mysql+mysqldb://justkaggle:"+"justkaggle1234!"+"@104.197.128.149/smy_db", encoding='utf-8')
#conn = engine.connect()

#ym_range = list(set(pd.date_range(start='2019-01',end='2019-12').strftime('%Y%m').tolist()))
year_range = ['2020']
month_range = ['01','02','03','04','05','06','07']

api_key = '5243515448677965343255436b7a49'
base_url = 'http://openapi.seoul.go.kr:8088/{}/json/CardBusTimeNew/{}/{}/{}/'

bus_ym_df = pd.DataFrame()
for j in year_range:
    start_time = time.time()
    table_name = 'bus_timelog_'+j
    
    for i in month_range:
        ym_value = j + i
        
        bus_ym_df0 = pd.DataFrame()
        num = 1
        while True:
            try:

                req = requests.get(base_url.format(api_key,str(num),str(num+999),ym_value), headers = {'user-agent': \
                                                ('Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Safari/537.36')})

                bus_json = req.json()
                sample_df = json_normalize(bus_json['CardBusTimeNew']['row'])

                #sample_df.to_sql(name=table_name,con=engine,if_exists='append')
                bus_ym_df0 = bus_ym_df0.append(sample_df)

                print('버스정류장 데이터 {} ~ {} row 수집중'.format(str(num),str(num+999)))
                num +=1000
            except:
                break
        print('--------------------'+j+'년'+i+'월 데이터 수집 완료---------------------')
        
        bus_ym_df = bus_ym_df.append(bus_ym_df0)
        
    print('============================================================================')
    print('=============================='+j+' 연도 테이블 생성 완료========================')
    print('============================================================================')
    print('연평균 수집 소요 시간(초) : ', time.time()-start_time)


/Users/sinmin-yong/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:29: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


버스정류장 데이터 1 ~ 1000 row 수집중
버스정류장 데이터 1001 ~ 2000 row 수집중
버스정류장 데이터 2001 ~ 3000 row 수집중
버스정류장 데이터 3001 ~ 4000 row 수집중
버스정류장 데이터 4001 ~ 5000 row 수집중
버스정류장 데이터 5001 ~ 6000 row 수집중
버스정류장 데이터 6001 ~ 7000 row 수집중
버스정류장 데이터 7001 ~ 8000 row 수집중
버스정류장 데이터 8001 ~ 9000 row 수집중
버스정류장 데이터 9001 ~ 10000 row 수집중
버스정류장 데이터 10001 ~ 11000 row 수집중
버스정류장 데이터 11001 ~ 12000 row 수집중
버스정류장 데이터 12001 ~ 13000 row 수집중
버스정류장 데이터 13001 ~ 14000 row 수집중
버스정류장 데이터 14001 ~ 15000 row 수집중
버스정류장 데이터 15001 ~ 16000 row 수집중
버스정류장 데이터 16001 ~ 17000 row 수집중
버스정류장 데이터 17001 ~ 18000 row 수집중
버스정류장 데이터 18001 ~ 19000 row 수집중
버스정류장 데이터 19001 ~ 20000 row 수집중
버스정류장 데이터 20001 ~ 21000 row 수집중
버스정류장 데이터 21001 ~ 22000 row 수집중
버스정류장 데이터 22001 ~ 23000 row 수집중
버스정류장 데이터 23001 ~ 24000 row 수집중
버스정류장 데이터 24001 ~ 25000 row 수집중
버스정류장 데이터 25001 ~ 26000 row 수집중
버스정류장 데이터 26001 ~ 27000 row 수집중
버스정류장 데이터 27001 ~ 28000 row 수집중
버스정류장 데이터 28001 ~ 29000 row 수집중
버스정류장 데이터 29001 ~ 30000 row 수집중
버스정류장 데이터 30001 ~ 31000 row 수집중
버스정류장 데이터 31001 ~ 32000 row 수집

# area code preprocessing

1. [지역 정보 코드](http://kssc.kostat.go.kr/ksscNew_web/kssc/common/CommonBoardList.do?gubun=1&strCategoryNameCode=019&strBbsId=kascrr&categoryMenu=014)

    a. 행정동: 기초자치단체의 하부행정기관으로 **행정구역, 주민센터**와 동일한 개념

    1) 행정기관코드: 행정동을 나타내는 코드 (ex. 서울:1100000000, 청운(효자)동:1111051500)

    2) 행정구역코드: 통계청에서 관리하는 코드 (ex. 서울:11, 종로구:11010, 청운(효자)동:1101072)

    → 분류체계만 다를 뿐, 관할 구역은 동일

    b. 법정동: 법으로 지정된 동이란 뜻으로 정부의 토지 구획, 지적도, 재산권 행사와 같은 법률 행위에서 일반적으로 사용되는 **주소지** 개념 (ex. 서울:1100000000, 청운(효자)동: 1111010100)

    → 행정동과 M:N 관계이며, 같은 이름이라도 구역이 조금씩 다를 수 있음

In [61]:
'''지역구 코드와 행정구역 코드 3,4번째 숫자가 같은지 검정'''
seoul = kor_address[kor_address['시도']=='서울특별시']
seoul2 = pd.merge(seoul,area_code,left_on='시군구',right_on='지역구',how='left')

for i in range(1,len(seoul2)):#서울제외
    if int(seoul2['행정구역코드'].iloc[i][2:4]) == int(seoul2['code'].iloc[i]):
        pass
    else:
        print(i)
print('valid finish')

valid finish


# 상권 데이터

# 버스 좌표 데이터 기반 시각화

- 3개월치의 데이터를 이용
- 정류소별 시간대별 평균을 낼 것

In [36]:
bus_coord = pd.read_csv('서울특별시 버스정류소 위치정보.csv',engine = 'python',encoding = 'cp949')

In [42]:
bus_coord['정류소번호'] = bus_coord['정류소번호'].apply(lambda x: re.sub('[^0-9]','',x))

In [49]:
bus07 = bus_ym_df[bus_ym_df['USE_MON'] == '202007']
bus06 = bus_ym_df[bus_ym_df['USE_MON'] == '202006']

In [52]:
bus06_fin = pd.merge(bus06,bus_coord,left_on = 'BSST_ARS_NO', right_on = '정류소번호', how = 'left')

In [61]:
# 정류소별 자정 승하차 인원
bus_df06 = bus06_fin[['BUS_STA_NM','BSST_ARS_NO','정류소번호','X좌표','Y좌표','MIDNIGHT_RIDE_NUM','MIDNIGHT_ALIGHT_NUM']]

In [67]:
# null값인 row확인
bus_df06.iloc[bus_df06.isnull()[bus_df06.isnull()['X좌표'] == True].index.tolist()]

BUS_STA_NM BSST_ARS_NO 정류소번호  X좌표  Y좌표  MIDNIGHT_RIDE_NUM  \
121    우이동도선사입구(기점가상)           ~   NaN  NaN  NaN                0.0   
203      동아운수종점(종점가상)           ~   NaN  NaN  NaN                0.0   
205        7단지영업소(가상)           ~   NaN  NaN  NaN                0.0   
270        7단지영업소(가상)           ~   NaN  NaN  NaN                0.0   
274          선일교통(가상)           ~   NaN  NaN  NaN                0.0   
...               ...         ...   ...  ...  ...                ...   
15764      7단지영업소(가상)           ~   NaN  NaN  NaN                0.0   
15771      7단지영업소(가상)           ~   NaN  NaN  NaN                8.0   
15773        철산리버빌아파트       38079   NaN  NaN  NaN                0.0   
15943        철산리버빌아파트       38003   NaN  NaN  NaN                0.0   
15956   양천공영차고지(기점가상)           ~   NaN  NaN  NaN                2.0   

       MIDNIGHT_ALIGHT_NUM  
121                    1.0  
203                    2.0  
205                    0.0  
270                    0.0  
274                    0.0  
...                    ...  
15764                  0.0  
15771                  0.0  
15773                  0.0  
15943                  0.0  
15956                  0.0  

[431 rows x 7 columns]

In [73]:
bus_df06.dropna(inplace = True)

/Users/sinmin-yong/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [59]:
# 서울시청 위도, 경도
latitude_value = 37.565954
longitude_value = 126.978023

map_hs = folium.Map((latitude_value, longitude_value), zoom_start = 15)

In [60]:
#승차, 하차 별로 그룹

mcg = folium.plugins.MarkerCluster(control=False)
bus_get_on = folium.plugins.FeatureGroupSubGroup(mcg,'bus_get_on')
bus_get_off = folium.plugins.FeatureGroupSubGroup(mcg,'bus_get_off')

In [74]:
for row in bus_df06.itertuples():
    x = float(row.X좌표)
    y = float(row.Y좌표)
    get_on = '승차: ' + str(row.MIDNIGHT_RIDE_NUM)
    get_off = '하차: ' + str(row.MIDNIGHT_ALIGHT_NUM)


    map_hs.add_child(mcg)
    map_hs.add_child(bus_get_on)
    map_hs.add_child(bus_get_off)

    get_on_icon=folium.Icon(color='green', icon='angle-up', icon_color="white", prefix='fa')
    get_off_icon=folium.Icon(color='red', icon='angle-down', icon_color="white", prefix='fa')

    bus_get_on.add_child(folium.Marker([y,x],icon=get_on_icon,popup = get_on))
    bus_get_off.add_child(folium.Marker([y,x],icon=get_off_icon,popup = get_off))

In [274]:
folium.LayerControl().add_to(map_hs)
map_hs.save('stations2.html')

# 지하철 좌표 데이터 기반 시각화

In [96]:
url = 'https://github.com/Just-Kaggle/labs/blob/master/1_part/1_Gis_dataset/%EC%84%9C%EC%9A%B8%EC%8B%9C%20%EC%97%AD%EC%BD%94%EB%93%9C%EB%A1%9C%20%EC%A7%80%ED%95%98%EC%B2%A0%EC%97%AD%20%EC%9C%84%EC%B9%98%20%EC%A1%B0%ED%9A%8C.csv'

req = requests.get(url)

soup = BeautifulSoup(req.content,'html.parser')
table = soup.find('table',{'class':'js-csv-data csv-data js-file-line-container'})

col = []
for i in table.find('thead').find_all('th'):
    
    col.append(i.get_text(strip = False))

tr_ls = table.find('tbody').find_all('tr')

value_ls = []
for i in tr_ls:
    td_ls = i.find_all('td')
    
    value_ls0 = []
    for idx,j in enumerate(td_ls):
        if idx == 0:
            continue
        else:
            td = j.get_text(strip = False)
            
            value_ls0.append(td)
            
        
    value_ls.append(value_ls0)
    
sub_addr = pd.DataFrame(value_ls,columns = col)

sub_addr2 = sub_addr.drop_duplicates()

In [165]:
sub_ym__df06 = sub_ym_df[sub_ym_df['USE_MON']=='202006'][['LINE_NUM','SUB_STA_NM','EIGHT_RIDE_NUM','EIGHT_ALIGHT_NUM']]

sub_ym__df07 = sub_ym_df[sub_ym_df['USE_MON']=='202007'][['LINE_NUM','SUB_STA_NM','EIGHT_RIDE_NUM','EIGHT_ALIGHT_NUM']]

In [131]:
sub_addr2['호선'].unique()

array(['8', '3', '5', '1', '4', '2', 'A', 'K', 'B', 'UI', 'G', '7', '6',
       'I', 'SU', 'U', 'S', 'KK', '9', 'I2', 'E'], dtype=object)

In [154]:
sub_ym__df07['LINE_NUM'].unique()

array(['1호선', '2호선', '5호선', '3호선', '6호선', '4호선', '7호선', '8호선', '9호선',
       '경강선', '경부선', '경원선', '경의중앙선', '경인선', '경춘선', '공항철도', '수인선', '과천선',
       '분당선', '안산선', '우이신설선', '장항선'], dtype=object)

In [ ]:
'''def sub_line(x):
    if x == 'A':
        return '공항철도'
    elif x == 'K':
        return '경의중앙선'
    elif x == 'B':
        return '분당선'
    elif x == 'UI':
        return '우이신설선'
    elif x == 'G':
        return '경춘선'
    elif x == 'I':
        return '인천1호선'
    elif x == 'I2':
        return '인천2호선'
    elif x == 'SU':
        return '수인선'
    elif x == 'U':
        return 
    
sub_ym__df07['LINE_NUM'] = sub_ym__df07['LINE_NUM'].\
                           apply(lambda x : '9호선' if x=='9호선2~3단계' else \
                                         ('공항철도' if x == '공항철도 1호선' else x))

sub_ym__df07['LINE_NUM'] = sub_ym__df07['LINE_NUM'].\
                           apply(lambda x : '경의중앙선' if x=='경의선' or x=='중앙선' else x)

sub_ym__df07['LINE_NUM'] = sub_ym__df07['LINE_NUM'].\
                           apply(lambda x : '3호선' if x=='일산선' or x=='중앙선' else x)'''

In [176]:
# 일단 예외 케이스가 너무 많기 때문에 호선 구분없이 지하철 역별 승하차 인원의 합을 sum 하기로 결정
# 그 다음 유동인구가 높은 역을 산정했을 때 해당 역의 환승 통로 및 어떤 환승구간에서 유동인구가 높은지 확인할 예정
sub_ym__df07_gb = sub_ym__df07.groupby('SUB_STA_NM').sum().reset_index()

#단순 중복제거이므로 가장 맨 위의 row가 채택되었을 것임
sub_addr3 = sub_addr2.drop_duplicates('전철역명')

In [175]:
subway07 = pd.merge(sub_ym__df07_gb,sub_addr3 , left_on = 'SUB_STA_NM', right_on = '전철역명', how = 'left')

In [198]:
coord_null = subway07.iloc[subway07.isnull()[subway07.isnull()['전철역명'] == True].index.tolist()].iloc[:,[0,1,2]]

coord_null['SUB_STA_NM'] = coord_null['SUB_STA_NM'].apply(lambda x : x.split('(')[0])
coord_null2 = pd.merge(coord_null,sub_addr3 , left_on = 'SUB_STA_NM', right_on = '전철역명', how = 'left')

subway07_notnull = subway07.iloc[subway07.isnull()[subway07.isnull()['전철역명'] == False].index.tolist()]

subway07_temp = pd.concat([subway07_notnull,coord_null2],axis = 0)
subway07_temp.index = range(len(subway07_temp))

In [280]:
subway07_temp2 = subway07_temp.iloc[subway07_temp.isnull()[subway07_temp.isnull()['전철역명'] == False].index.tolist()]

In [282]:
subway07_temp2.rename(columns = {'X좌표(WGS)' : 'y_coord','Y좌표(WGS)' : 'x_coord'},inplace = True)

/Users/sinmin-yong/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [283]:
subway07_temp3 = subway07_temp2[subway07_temp2['x_coord'] != '']

subway07_temp3['x_coord'] = subway07_temp3['x_coord'].astype('float64')
subway07_temp3['y_coord'] = subway07_temp3['y_coord'].astype('float64')

/Users/sinmin-yong/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/Users/sinmin-yong/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [284]:
subway07_temp3.index = range(len(subway07_temp3))

In [290]:
# 서울시청 위도, 경도
latitude_value = 37.565954
longitude_value = 126.978023

map_hs = folium.Map((latitude_value, longitude_value), zoom_start = 15)

In [291]:
#승차, 하차 별로 그룹

mcg = folium.plugins.MarkerCluster(control=False)
subway_get_on = folium.plugins.FeatureGroupSubGroup(mcg,'subway_get_on')
subway_get_off = folium.plugins.FeatureGroupSubGroup(mcg,'subway_get_off')

In [292]:
for row in subway07_temp3.itertuples():
    x = float(row.x_coord)
    y = float(row.y_coord)
    get_on = '승차: ' + str(row.EIGHT_RIDE_NUM)
    get_off = '하차: ' + str(row.EIGHT_ALIGHT_NUM)


    map_hs.add_child(mcg)
    map_hs.add_child(subway_get_on)
    map_hs.add_child(subway_get_off)

    get_on_icon=folium.Icon(color='green', icon='angle-up', icon_color="white", prefix='fa')
    get_off_icon=folium.Icon(color='red', icon='angle-down', icon_color="white", prefix='fa')

    subway_get_on.add_child(folium.Marker([y,x],icon=get_on_icon,popup = get_on))
    subway_get_off.add_child(folium.Marker([y,x],icon=get_off_icon,popup = get_off))

In [293]:
folium.LayerControl().add_to(map_hs)
map_hs.save('subway_stations.html')

In [294]:
subway07_temp3

SUB_STA_NM  EIGHT_RIDE_NUM  EIGHT_ALIGHT_NUM 전철역코드     전철역명 호선  외부코드  \
0           가능         21490.0           15130.0  1907       가능  1   109   
1         가락시장         36651.0           71822.0  2818     가락시장  8   817   
2      가산디지털단지         36557.0          572780.0  2748  가산디지털단지  7   746   
3           가양         74696.0           41887.0  4107       가양  9   907   
4           가좌         24749.0            8081.0  1265       가좌  K  K315   
..         ...             ...               ...   ...      ... ..   ...   
482        충정로         30403.0           98523.0  2532      충정로  5   531   
483      한성대입구         39935.0           28703.0  0419    한성대입구  4   419   
484        화랑대         56482.0           12844.0  2647      화랑대  6   646   
485         회현          7319.0          132902.0  0425       회현  4   425   
486         흑석         21877.0           14978.0  4119       흑석  9   919   

    사이버스테이션     X좌표      Y좌표    y_coord     x_coord  
0      1907  509550  1179510  37.748577  127.044213  
1      2818  525992  1108579  37.492522  127.118234  
2      1702  473861  1105307  37.481072  126.882343  
3      4107  467700  1127500  37.561391  126.854456  
4      1265  480980  1129579  37.568491  126.915487  
..      ...     ...      ...        ...         ...  
482    0243  491842  1127227  37.559973  126.963672  
483    0419  501130  1135052  37.588458  127.006221  
484    2647  518457  1143835  37.620064  127.084689  
485    0425  495072  1126787  37.558514  126.978246  
486    4119  491832  1112978  37.508770  126.963708  

[487 rows x 12 columns]

In [295]:
bus06

USE_MON BUS_ROUTE_NO          BUS_ROUTE_NM STND_BSST_ID BSST_ARS_NO  \
0    202006          100        100번(하계동~용산구청)    110000327       11428   
1    202006          100        100번(하계동~용산구청)    110000335       11436   
2    202006          100        100번(하계동~용산구청)    107000012       08102   
3    202006          100        100번(하계동~용산구청)    107000013       08103   
4    202006          100        100번(하계동~용산구청)    107000014       08104   
..      ...          ...                   ...          ...         ...   
995  202006          N62  N62번(양천공영차고지~면목동차고지)    114000219       15323   
996  202006          N62  N62번(양천공영차고지~면목동차고지)    114000221       15325   
997  202006          N62  N62번(양천공영차고지~면목동차고지)    114000225       15329   
998  202006          N62  N62번(양천공영차고지~면목동차고지)    114000250       15354   
999  202006          N62  N62번(양천공영차고지~면목동차고지)    114000261       15366   

         BUS_STA_NM  MIDNIGHT_RIDE_NUM  MIDNIGHT_ALIGHT_NUM  ONE_RIDE_NUM  \
0            한성여객종점                0.0                  5.0           0.0   
1          하계1동주민센터                0.0                  0.0           0.0   
2              창문여고              159.0                251.0          20.0   
3              동방고개               17.0                123.0           5.0   
4         장위동그린빌아파트               12.0                139.0           1.0   
..              ...                ...                  ...           ...   
995      푸른마을2단지아파트               22.0                  1.0          14.0   
996  신월시영아파트.강월초등학교               18.0                  0.0          21.0   
997         금실어린이공원               29.0                  1.0          14.0   
998       신월시영아파트중문               20.0                  1.0           9.0   
999        신월문화체육센터               15.0                  0.0          18.0   

     ONE_ALIGHT_NUM  ...  NINETEEN_ALIGHT_NUM  TWENTY_RIDE_NUM  \
0               2.0  ...                  4.0              6.0   
1               0.0  ...                 19.0            221.0   
2              41.0  ...                532.0            524.0   
3              15.0  ...                187.0             37.0   
4              18.0  ...                459.0              9.0   
..              ...  ...                  ...              ...   
995             1.0  ...                  0.0              0.0   
996             0.0  ...                  0.0              0.0   
997             2.0  ...                  0.0              0.0   
998             0.0  ...                  0.0              0.0   
999             2.0  ...                  0.0              0.0   

     TWENTY_ALIGHT_NUM  TWENTY_ONE_RIDE_NUM  TWENTY_ONE_ALIGHT_NUM  \
0                  3.0                 19.0                    6.0   
1                 17.0                253.0                   17.0   
2                473.0                456.0                  518.0   
3                186.0                 48.0                  194.0   
4                342.0                 12.0                  307.0   
..                 ...                  ...                    ...   
995                0.0                  0.0                    0.0   
996                0.0                  0.0                    0.0   
997                0.0                  0.0                    0.0   
998                0.0                  0.0                    0.0   
999                0.0                  0.0                    0.0   

     TWENTY_TWO_RIDE_NUM  TWENTY_TWO_ALIGHT_NUM  TWENTY_THREE_RIDE_NUM  \
0                    5.0                    6.0                    0.0   
1                  276.0                    4.0                    0.0   
2                  462.0                  452.0                  229.0   
3                   54.0                  162.0                   12.0   
4                   11.0                  288.0                    8.0   
..                   ...                    ...                    ...   
995                  0.0  

# 

- 스마트도서관에 연동되어 있는 도서관
- 역 내에 설치할 수 있는 공간이 있는지
    - 설치의 현실성


# TO DO
- 지금까지의 데이터셋 정리 작업 (민걸)
    - 유동인구 데이터 - 최근 1달
    - 인구분포 데이터 - 최근 1달
    - 서울 내 도서관 위치 정보
    - 서울 내 스마트 도서관 위치 정보
    - 인기도서 - 최소단위가 구 / 책 단위로 접근x -> 장르별로 접근 - 1달 단위 1년
    - 버스, 지하철 월별 시간별 - 최근 1달
- 서울시 대출 데이터 수집 (필령)
- 코로나 데이터 수집 (민용)
- 시각화 좀 해놓을게요(민용, 민걸)
-------> 화요일까지 하는걸로


# folium 활용 시각화

In [310]:
subway07_temp3[['SUB_STA_NM','x_coord','y_coord','EIGHT_RIDE_NUM','EIGHT_ALIGHT_NUM']].to_csv('subway07_tableau.csv',encoding = 'utf8')

In [312]:
subway07_temp3

SUB_STA_NM  EIGHT_RIDE_NUM  EIGHT_ALIGHT_NUM 전철역코드     전철역명 호선  외부코드  \
0           가능         21490.0           15130.0  1907       가능  1   109   
1         가락시장         36651.0           71822.0  2818     가락시장  8   817   
2      가산디지털단지         36557.0          572780.0  2748  가산디지털단지  7   746   
3           가양         74696.0           41887.0  4107       가양  9   907   
4           가좌         24749.0            8081.0  1265       가좌  K  K315   
..         ...             ...               ...   ...      ... ..   ...   
482        충정로         30403.0           98523.0  2532      충정로  5   531   
483      한성대입구         39935.0           28703.0  0419    한성대입구  4   419   
484        화랑대         56482.0           12844.0  2647      화랑대  6   646   
485         회현          7319.0          132902.0  0425       회현  4   425   
486         흑석         21877.0           14978.0  4119       흑석  9   919   

    사이버스테이션     X좌표      Y좌표    y_coord     x_coord  
0      1907  509550  1179510  37.748577  127.044213  
1      2818  525992  1108579  37.492522  127.118234  
2      1702  473861  1105307  37.481072  126.882343  
3      4107  467700  1127500  37.561391  126.854456  
4      1265  480980  1129579  37.568491  126.915487  
..      ...     ...      ...        ...         ...  
482    0243  491842  1127227  37.559973  126.963672  
483    0419  501130  1135052  37.588458  127.006221  
484    2647  518457  1143835  37.620064  127.084689  
485    0425  495072  1126787  37.558514  126.978246  
486    4119  491832  1112978  37.508770  126.963708  

[487 rows x 12 columns]

In [311]:
with pd.ExcelWriter('output.xlsx') as writer:  
    subway07_temp3[['SUB_STA_NM','x_coord','y_coord','EIGHT_RIDE_NUM','EIGHT_ALIGHT_NUM']].to_excel(writer, sheet_name='Sheet_name_1')

In [297]:
# 서울시청 위도, 경도
latitude_value = 37.565954
longitude_value = 126.978023

center = [37.565954,126.978023]
map_hs = folium.Map(location = center, zoom_start = 15)

In [ ]:
json_data = 

In [302]:
subway07_temp3[['SUB_STA_NM','x_coord','y_coord','EIGHT_RIDE_NUM','EIGHT_ALIGHT_NUM']]

SUB_STA_NM     x_coord    y_coord  EIGHT_RIDE_NUM  EIGHT_ALIGHT_NUM
0           가능  127.044213  37.748577         21490.0           15130.0
1         가락시장  127.118234  37.492522         36651.0           71822.0
2      가산디지털단지  126.882343  37.481072         36557.0          572780.0
3           가양  126.854456  37.561391         74696.0           41887.0
4           가좌  126.915487  37.568491         24749.0            8081.0
..         ...         ...        ...             ...               ...
482        충정로  126.963672  37.559973         30403.0           98523.0
483      한성대입구  127.006221  37.588458         39935.0           28703.0
484        화랑대  127.084689  37.620064         56482.0           12844.0
485         회현  126.978246  37.558514          7319.0          132902.0
486         흑석  126.963708  37.508770         21877.0           14978.0

[487 rows x 5 columns]

# 데이터 활용 방안

1. 유동인구 및 인구분포 데이터
    - 지수화 용도
2. 서울내 도서관 위치정보, 스마트도서관 위치정보
3. 인기도서
4. 도서관별 대출
    - 코로나 전 후 대출 시각화
5. 버스 승하차, 지하철 승하차

idea
ex) 광화문에 도서관이 있다고 친다 -> 직장인 타겟 -> 교통 편리한 곳
ex) 강서구 -> 학생 타겟 ->


# 분석 플로우
1. 왜 우리가 해당 주제를 잡고 분석을 진행했는지?
    - 코로나 (covid19)
    - 도서관 이용률(필령형 데이터)
    
2. 분석
    2-1. 현 스마트 도서관 위치
        - 현 스마트 도서관 위치의 문제점 / 아쉬운 점
        - 특정 도서관의 이용률
    2-2. 
    
3. 본론(최적 위치 선정 과정)
    3-1. 지역별 수집한 데이터 시각화 및 분석 자료
        - 교통, 인구, 지하철 등...
    3-2. 지수 개발 과정
    3-3. 설치의 현실성
    3-4. 지역별 인기도서 리스트 바탕으로 효율적인 책 배치
        - 지역적 특색파악
        - 장르 파악
        - 시각화 : 지역별 워드클라우드
        - 스마트도서관 워드클라우드를 띄어주는 것
            - 키워드 클릭하면 그것과 관련한 책 추천

4. 결론 및 시사점
    4-1. 효과
        - 우리가 지수 개발 -> 효과를 어떻게 측정할 수 있을지? : 해당 지수를 바탕으로 위치 선정 -> 주변 도서관 확인
        - 인구밀도도 고려하지만 지역적 특색을 파악하여 지역주민 및 유동인구가 만족감 부여 가능
    4-2. 한계점 및 제언
        - 필요한 데이터 나열하고 -> 더 고도화된 스마트 도서관을 제공 가능


# 민걸
- 데이터 정리
- 데이터 전처리 -> 시각화(태블로)
- 지수

# 필령
- 대출 데이터와 코로나 데이터 비교분석(1)
- 지수

# 민용
- 인기도서 기반 자료정리 (3-4)
- 지수
- 토요일 방알아보기



In [ ]:

# 지수 개발
- 논문 인용 + 우리가 개발
- 자료조사 필요
